### PR TITLE
Allow Symfony 6

### PR DESCRIPTION
**Pull request by Kocal**
- Dropped Symfony 3.4, 5.0, 5.1 and 5.2 support
- Added Symfony 6 support

### CODE REVIEW

**Comment by stof** [comment]
Careful about that, as we should avoid _releasing_ a version advocating support for Symfony 6 until the RC releases IMO (as the list of BC breaks is not final yet)

**Comment by dreadnip** [comment]
Friendly reminder :slightly_smiling_face: Anything holding this back from being merged?

File: .travis.yml
Comment (dmaicher) [inline .travis.yml:17]:
```suggestion
        - SYMFONY_VERSION='6.0.*'
```

?

Comment (Kocal) [reply .travis.yml:17]: Yep, now I can change to `*` since Symfony 6 has been released :+1:

**Comment by Kocal** [comment]
Hum, it looks like I can't type methods to make Symfony 6 happy without breaking Symfony <= 4.. should we drop support of Symfony 3.3 and 4?

**Comment by thePanz** [comment]
@Kocal how is the DoctrineBundle doing in that case? it still supports those versions of SF :thinking: 
see: https://github.com/doctrine/DoctrineBundle/blob/2.6.x/composer.json

**Comment by Kocal** [comment]
@thePanz AFAIK they don't have to implement a `SerializerInterface` or any interfaces that changed between SF 3.x and 6.x.

Before a223ee7263ae0e5f2baa6a76cfc356269ed48250, I have the following error with Symfony 6 build:
```
Declaration of FOS\JsRoutingBundle\Serializer\Normalizer\RoutesResponseNormalizer::normalize($data, $format = null, array $context = []) must be compatible with Symfony\Component\Serializer\Normalizer\NormalizerInterface::normalize(mixed $object, ?string $format = null, array $context = []): ArrayObject|array|string|int|float|bool|null in /home/travis/build/FriendsOfSymfony/FOSJsRoutingBundle/Serializer/Normalizer/RoutesResponseNormalizer.php on line 25
```

And after a223ee7263ae0e5f2baa6a76cfc356269ed48250, I have the following error with Symfony 3.4/4.4/5.x builds:
 ```
Declaration of FOS\JsRoutingBundle\Serializer\Normalizer\RoutesResponseNormalizer::normalize($data, ?string $format = NULL, array $context = Array): array must be compatible with Symfony\Component\Serializer\Normalizer\NormalizerInterface::normalize($object, $format = NULL, array $context = Array) in /home/travis/build/FriendsOfSymfony/FOSJsRoutingBundle/Serializer/Normalizer/RoutesResponseNormalizer.php on line 20
```

Maybe the solution is to create a custom `NormalizerInterface` that will be dynamic in function of the Symfony version?

**Comment by dmaicher** [comment]
Actually Serializer return types have been relaxed after releasing 6.0 [here](https://github.com/symfony/symfony/pull/44331).

So this will be part of 6.0.1.

Can you test if you can make it work with latest 6.0.x-dev?

File: .travis.yml
Comment (Kocal) [reply .travis.yml:16]: I've re-added `6.0.x-dev` until Symfony 6.0.1 is released.

**Comment by sebheitzmann** [comment]
can you allow also PHP8.1 ?

**Comment by mhpcc** [comment]
php 8.1 is allowed (^8.1), it just has to be added to travis

**Comment by sebheitzmann** [comment]
> php 8.1 is allowed (^8.0), it just has to be added to travis

Your right, i'm sorry. Any date for the sf6 support ?

**Comment by Chris53897** [comment]
Maybe a similar solution to this is possible for symfony 4.4 support. https://github.com/briannesbitt/Carbon/issues/2508#issuecomment-983756110

In my eyes a new major version php8 only with support for symfony 5.4 and 6.x should be created to lower the burden of the maintainers.

**Comment by sebheitzmann** [comment]
Could you create a 3.0 branch with a sf6 support ? This bundle is the last one blocking my project upgrade.

**Comment by dreadnip** [comment]
@Kocal The two remaining errors:
```
Fatal error:  Declaration of FOS\JsRoutingBundle\Serializer\Denormalizer\RouteCollectionDenormalizer::supportsDenormalization($data, string $type, ?string $format = NULL): bool must be compatible with Symfony\Component\Serializer\Normalizer\DenormalizerInterface::supportsDenormalization($data, $type, $format = NULL) in /home/travis/build/FriendsOfSymfony/FOSJsRoutingBundle/Serializer/Denormalizer/RouteCollectionDenormalizer.php on line 18

Fatal error: Declaration of FOS\JsRoutingBundle\Serializer\Denormalizer\RouteCollectionDenormalizer::supportsDenormalization($data, string $type, ?string $format = NULL): bool must be compatible with Symfony\Component\Serializer\Normalizer\DenormalizerInterface::supportsDenormalization($data, $type, $format = NULL) in /home/travis/build/FriendsOfSymfony/FOSJsRoutingBundle/Serializer/Denormalizer/RouteCollectionDenormalizer.php on line 18```

seem to be related to [this change](https://github.com/FriendsOfSymfony/FOSJsRoutingBundle/pull/408/commits/a223ee7263ae0e5f2baa6a76cfc356269ed48250#diff-2fd84bc6297fb28807adaeb18dde6a5020f8f8290885ba16846a57b8d01b59beR41) you introduced in a223ee7263ae0e5f2baa6a76cfc356269ed48250. You've reverted some of these changes in 40108d5395c8b4f380ced83d9bafb71ea5fe5210, but not the supportsDenormalization bool return type. Could you also revert the return type on that method? It is also reverted in the (Symfony PR that eased up the return types)[https://github.com/symfony/symfony/pull/44331/files#diff-a9feffc12ca8263b6a76504af0aaa93f366eb57978aa1989621a19227e25da6eR353]. Afterwards, we wait for Symfony 6.0.1 I guess :slightly_smiling_face:

**Comment by Bukashk0zzz** [comment]
And Symfony 6.0.1 is here :) Any updates?

**Comment by mhpcc** [comment]
symfony 6.0.1 has not removed the return type declarations from serializer

**Comment by tobias-93** [comment]
Hi,

Sorry, this bundle has been under my radar for a long time.
Please rebase on the newest master to test this on all dependencies and PHP versions we're currently supporting (using Github Actions).
I'll review it when you've done this so we can finally support SF6.

**Comment by Kocal** [comment]
Hi @tobias-93, no worries! :) 

The PR has been rebased on latest master. I've removed Symfony 3.4 and 5.0/5.1/5.2 support since those versions are not maintened anymore.

I would still need help with normalizers params/returns types conflicts.

File: Serializer/Normalizer/RouteCollectionNormalizer.php
Comment (tobias-93) [inline Serializer/Normalizer/RouteCollectionNormalizer.php:26]:
```suggestion
    public function normalize($data, string $format = null, array $context = array())
```

File: Serializer/Normalizer/RoutesResponseNormalizer.php
Comment (tobias-93) [inline Serializer/Normalizer/RoutesResponseNormalizer.php:26]:
```suggestion
    public function normalize($data, string $format = null, array $context = array())
```

File: composer.json
Comment (tobias-93) [inline composer.json:21]:
```suggestion
        "symfony/serializer": "^4.4.20|^5.3|^6.0.1",
```

**Review by tobias-93 — CHANGES_REQUESTED** [review]
I think this should fix the forward compatibility given the Symfony Bugfix as mentioned by @dreadnip

**Comment by Kocal** [comment]
The CI is green, I've literally removed types on (de)normalizers... :grimacing: 

I think we can add them back when dropping support of Symfony 4.4 or/and 5.3.

**Comment by sebheitzmann** [comment]
Good news. Congrat

File: Serializer/Denormalizer/RouteCollectionDenormalizer.php
Comment (dmaicher) [inline Serializer/Denormalizer/RouteCollectionDenormalizer.php:23]: This might be a BC break. I think we should remove the return type?

```suggestion
    public function denormalize($data, $type, $format = null, array $context = [])
```

File: Serializer/Denormalizer/RouteCollectionDenormalizer.php
Comment (dmaicher) [inline Serializer/Denormalizer/RouteCollectionDenormalizer.php:22]: should keep this one to avoid the reported deprecation notice

**Comment by stof** [comment]
The way to be compatible with all versions is to support only PHP 7.2+ and to add the return type in your class (but not the argument type). This way, you are compatible with both versions thanks to variance rules.

**Comment by tobias-93** [comment]
Thanks @Kocal

**Pull request merged**
